### PR TITLE
Invariants Improvement: Add multiple pool support in position and rewards manager invariant testing

### PR DESCRIPTION
**Pull request by prat-gpt**
# Description of change
### Contracts
* In rewardsManager, `rewardsClaimed `and `updateRewardsClaimed` were stored per epoch https://github.com/ajna-finance/contracts/blob/develop/src/RewardsManager.sol#L71. Since we can have multiple pools so rewardsClaimed should be stored per pool per epoch rather than only per epoch.
* Fixed with storing `rewardsClaimed `and `updateRewardsClaimed` per pool per epoch.

### Invariants 
* Updated Position and rewards Manager invariants to test against multiple pools.
* Add configurable `NO_OF_POOLS` env var for position and rewards manager

### CODE REVIEW

**Review by grandizzy — APPROVED** [review]
Great catch, looks good, I think ideally instead all those mappings we could have only 2 `pool address` -> `PoolRewardsInfo` and `token id` -> `StakeInfo` e.g.

```Solidity
    /// @dev Pool address `pool => Pool Rewards info`.
    mapping(address => PoolRewardsInfo) internal poolRewardsInfo;
    /// @dev Mapping `tokenID => Stake info`.
    mapping(uint256 => StakeInfo) internal stakes;

    struct PoolRewardsInfo {
        mapping(uint256 => uint256) rewardsClaimed;
        mapping(uint256 => uint256) updateBucketRewardsClaimed;
        mapping(address => mapping(uint256 => mapping(uint256 => uint256))) bucketExchangeRates;
    }

    struct StakeInfo {
        address ajnaPool;                         // address of the Ajna pool the NFT corresponds to
        uint96  lastClaimedEpoch;                 // last epoch the stake claimed rewards
        address owner;                            // owner of the LP NFT
        uint96  stakingEpoch;                     // epoch at staking time
        mapping(uint256 => BucketState) snapshot; // the LP NFT's balances and exchange rates in each bucket at the time of staking
        mapping(uint256 => bool) isEpochClaimed;
    }
```

File: tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol [hunk 1]
```diff
@@ -68,4 +74,23 @@ contract ERC721PoolRewardsHandler is RewardsPoolHandler, ReserveERC721PoolHandle
             if (indexes_.length != 0) { _updateExchangeRate(indexes_); }
         }
     }
+
+    modifier useRandomPool(uint256 poolIndex) override {
```
Comment (MikeHathaway) [inline tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol:78]: May be simpler to use an `EnnumerableSet` for the pools tracking, but this works for now

File: tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol [hunk 2]
```diff
@@ -13,14 +17,16 @@ contract ERC721PoolRewardsHandler is RewardsPoolHandler, ReserveERC721PoolHandle
     constructor(
         address rewards_,
         address positions_,
-        address pool_,
+        PoolInfo[10] memory pools_,
```
Comment (MikeHathaway) [inline tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol]: Would be good to parameterize the number of pools with an environment variable

**Review by MikeHathaway — APPROVED** [review]
LGTM! Few comments about potential improvements to the invariant tests but happy to leave out of scope here

File: tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol (hunk 2 above)
Comment (prat-gpt) [reply tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol]: Added in [a4310b2](https://github.com/ajna-finance/contracts/pull/927/commits/a4310b2495e6a823224c5ff4c5cfe29d477b90a2)

File: tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol (hunk 1 above)
Comment (prat-gpt) [reply tests/forge/invariants/PositionsAndRewards/handlers/ERC721PoolRewardsHandler.sol:78]: Updated array of PoolInfo struct to dynamic array of pool address in [a4310b2](https://github.com/ajna-finance/contracts/pull/927/commits/a4310b2495e6a823224c5ff4c5cfe29d477b90a2)

File: tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol [hunk 3]
```diff
@@ -24,12 +24,10 @@ abstract contract BaseERC20PoolHandler is BaseHandler {
     constructor(
         address pool_,
         address ajna_,
-        address quote_,
-        address collateral_,
```
Comment (EdNoepel) [inline tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol:28]: Nice cleanup.

File: tests/forge/invariants/PositionsAndRewards/ERC20PoolPositionsInvariants.t.sol [hunk 4]
```diff
@@ -21,39 +23,43 @@ import { PositionsInvariants }         from './PositionsInvariants.sol';
...
         _erc721impl        = _erc721poolFactory.implementation();
-        _erc20pool         = ERC20Pool(_erc20poolFactory.deployPool(address(_collateral), address(_quote), 0.05 * 10**18));
-        _pool              = Pool(address(_erc20pool));
         _positionManager   = new PositionManager(_erc20poolFactory, _erc721poolFactory);
 
+        uint256 noOfPools = vm.envOr("NO_OF_POOLS", uint256(10));
+
+        for (uint256 i = 0; i < noOfPools; ++i) {
+            address collateral = address(new TokenWithNDecimals(string(abi.encodePacked("Collateral", Strings.toString(i + 1))), "C", uint8(vm.envOr("COLLATERAL_PRECISION", uint256(18)))));
+            address quote      = address(new TokenWithNDecimals(string(abi.encodePacked("Quote", Strings.toString(i + 1))), "Q", uint8(vm.envOr("QUOTE_PRECISION", uint256(18)))));
```
Comment (EdNoepel) [inline tests/forge/invariants/PositionsAndRewards/ERC20PoolPositionsInvariants.t.sol:41]: Happy to see these env vars honored.

**Review by EdNoepel — APPROVED** [review]
Looks great, and tests run clean on `parvati`.

**Review by mattcushman — APPROVED** [review]
lgtm

**Review by dmitriia — COMMENTED** [review]
Looks ok, just a couple of info/gas comments:

A SLOAD can be saved and there no need to write if cap was reached:

https://github.com/ajna-finance/contracts/blob/a4310b2495e6a823224c5ff4c5cfe29d477b90a2/src/RewardsManager.sol#L650-L659

```diff
                uint256 rewardsClaimedInEpoch = updateRewardsClaimed[curBurnEpoch];

                // update total tokens claimed for updating bucket exchange rates tracker
                if (rewardsClaimedInEpoch + updatedRewards_ >= rewardsCap) {
                    // if update reward is greater than cap, set to remaining difference
-                   updatedRewards_ = rewardsClaimedInEpoch > rewardsCap ? 0 : rewardsCap - rewardsClaimedInEpoch;
+                   if (rewardsClaimedInEpoch >= rewardsCap) return;
+                   unchecked { updatedRewards_ = rewardsCap - rewardsClaimedInEpoch; }
                }

                // accumulate the full amount of additional rewards
-               updateRewardsClaimed[curBurnEpoch] += updatedRewards_;
+               updateRewardsClaimed[curBurnEpoch] = rewardsClaimedInEpoch + updatedRewards_;
```

Also, shouldn't `@dev` descriptions of the mappings moved be copied over from the initial structures?

https://github.com/sherlock-audit/2023-04-ajna-dmitriia/blob/073a388510c1adba9c629d59813f47e79156f8c7/ajna-core/src/RewardsManager.sol#L70-L76

```solidity
    /// @dev `epoch => rewards claimed` mapping.
    mapping(uint256 => uint256) public override rewardsClaimed;
    /// @dev `epoch => update bucket rate rewards claimed` mapping. Tracks the total amount of update rewards claimed.
    mapping(uint256 => uint256) public override updateRewardsClaimed;

    /// @dev Mapping of per pool bucket exchange rates at a given burn event `poolAddress => bucketIndex => epoch => bucket exchange rate`.
    mapping(address => mapping(uint256 => mapping(uint256 => uint256))) internal bucketExchangeRates;
```

Now it is rather obscure:

https://github.com/ajna-finance/contracts/blob/a4310b2495e6a823224c5ff4c5cfe29d477b90a2/src/interfaces/rewards/IRewardsManagerState.sol#L101-L105

```solidity
struct PoolRewardsInfo {
    mapping(uint256 => uint256) rewardsClaimed;                           // staking rewards claimed in a epoch
    mapping(uint256 => uint256) updateBucketRewardsClaimed;               // update exchange rate rewards claimed in a epoch
    mapping(uint256 => mapping(uint256 => uint256)) bucketExchangeRates;  // bucket exchange rates at a given burn event for each bucket
}

**Comment by prat-gpt** [comment]
@dmitriia Added gas improvements as per your comment here https://github.com/ajna-finance/contracts/pull/929. I have not updated the comments for consistency in the rest of the code.

**Comment by dmitriia** [comment]
Looks ok

**Pull request merged**